src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,620 @@
+                }
+            }
+
+            // We need to filter if a non-empty strict subset of filters are selected
+            var selectedFilters = data.SelectedFilters.Where(f => f.IsSelected).Select(f => f.Filter).ToImmutableArray();
+            var needToFilter = selectedFilters.Length > 0 && selectedFilters.Length < data.SelectedFilters.Length;
+
+            var initialListOfItemsToBeIncluded = new List<FilterResult>();
+            foreach (var item in data.InitialSortedList)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (needToFilter && ShouldBeFilteredOutOfCompletionList(item, selectedFilters))
+                {
+                    continue;
+                }
+
+                if (MatchesFilterText(item.FilterText, item.DisplayText, GetMatchPriority(item), filterText, data.Trigger.Reason))
+                {
+                    initialListOfItemsToBeIncluded.Add(new FilterResult(item, filterText, matchedFilterText: true));
+                }
+                else
+                {
+                    if (data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Deletion ||
+                        data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Backspace ||
+                        data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Invoke ||
+                        filterText.Length <= 1)
+                    {
+                        initialListOfItemsToBeIncluded.Add(new FilterResult(item, filterText, matchedFilterText: false));
+                    }
+                }
+            }
+
+            var snapshotForDocument = data.InitialSortedList
+                .FirstOrDefault(i => i.Properties.ContainsProperty(CompletionItemSource.TriggerBuffer))?.
+                Properties.GetProperty<ITextBuffer>(CompletionItemSource.TriggerBuffer).CurrentSnapshot ?? data.Snapshot;
+
+            if (initialListOfItemsToBeIncluded.Count == 0)
+            {
+                var document = snapshotForDocument.GetOpenDocumentInCurrentContextWithChanges();
+                var completionService = document.GetLanguageService<CompletionService>();
+                var completionRules = completionService.GetRules();
+                return HandleAllItemsFilteredOut(data.Trigger.Reason, data.SelectedFilters, selectedFilters, completionRules);
+            }
+
+            // If this was deletion, then we control the entire behavior of deletion ourselves.
+            if (data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Deletion || data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Backspace)
+            {
+                return HandleDeletionTrigger(data.InitialSortedList, data.Trigger.Reason, data.SelectedFilters, data.Trigger.Reason, filterText, initialListOfItemsToBeIncluded);
+            }
+
+            var caretPoint = session.TextView.GetCaretPoint(data.Snapshot.TextBuffer);
```

Comment:
This uses the buffer graph which isn't safe on a background thread. At least, it has a lot of mutable fields that aren't locked. @olegtk or @AmadeusW what's the take here?

Reply:
@amwieczo@microsoft.com , @olegtk, please help

---
In reply to: [221394741](https://github.com/dotnet/roslyn/pull/29016#discussion_r221394741) [](ancestors = 221394741)